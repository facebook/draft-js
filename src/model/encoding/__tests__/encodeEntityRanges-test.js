/**
 * Copyright (c) 2013-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @emails oncall+ui_infra
 * @format
 */

'use strict';

jest.disableAutomock();

const ContentBlock = require('ContentBlock');
const Immutable = require('immutable');

const createCharacterList = require('createCharacterList');
const encodeEntityRanges = require('encodeEntityRanges');

const {NONE} = require('DraftEntitySet');

const {OrderedSet, Repeat} = Immutable;

const SIX = OrderedSet.of('6');
const EIGHT = OrderedSet.of('8');

const createBlock = (text, entities) => {
  const style = OrderedSet();
  return new ContentBlock({
    key: 'a',
    text,
    type: 'unstyled',
    characterList: createCharacterList(
      Repeat(style, text.length).toArray(),
      entities,
    ),
  });
};

test('must return an empty array if no entities present', () => {
  const block = createBlock(' '.repeat(20), Repeat(NONE, 20).toArray());
  let encoded = encodeEntityRanges(block, {});
  expect(encoded).toMatchSnapshot();

  encoded = encodeEntityRanges(block, {'0': '0'});
  expect(encoded).toMatchSnapshot();
});

test('must return ranges with the storage-mapped key', () => {
  const entities = [NONE, NONE, SIX, SIX, NONE];
  const block = createBlock(' '.repeat(entities.length), entities);
  const encoded = encodeEntityRanges(block, {_6: '0'});
  expect(encoded).toMatchSnapshot();
});

test('must return ranges with multiple entities present', () => {
  const entities = [NONE, NONE, SIX, SIX, NONE, EIGHT, EIGHT, NONE];
  const block = createBlock(' '.repeat(entities.length), entities);
  const encoded = encodeEntityRanges(block, {_6: '0', _8: '1'});
  expect(encoded).toMatchSnapshot();
});

it('must return overlapping ranges with multiple entities present', () => {
  var entities = [NONE, NONE, SIX, SIX, SIX.union(EIGHT), EIGHT, EIGHT, NONE];
  var block = createBlock(' '.repeat(entities.length), entities);
  var encoded = encodeEntityRanges(block, {_6: '0', _8: '1'});
  expect(encoded).toMatchSnapshot();
});

it('must return interior ranges with multiple entities present', () => {
  var entities = [NONE, NONE, SIX, SIX, SIX.union(EIGHT), SIX, SIX, NONE];
  var block = createBlock(' '.repeat(entities.length), entities);
  var encoded = encodeEntityRanges(block, {_6: '0', _8: '1'});
  expect(encoded).toMatchSnapshot();
});

test('must return ranges with an entity present more than once', () => {
  const entities = [NONE, NONE, SIX, SIX, NONE, SIX, SIX, NONE];
  const block = createBlock(' '.repeat(entities.length), entities);
  const encoded = encodeEntityRanges(block, {_6: '0', _8: '1'});
  expect(encoded).toMatchSnapshot();
});

test('must handle ranges that include surrogate pairs', () => {
  const str = 'Take a \uD83D\uDCF7 #selfie';
  // prettier-ignore
  const entities = [
      NONE, NONE, NONE, NONE, NONE, NONE, // `Take a`
      SIX, SIX, SIX, SIX, SIX, SIX,       // ` [camera] #s`
      NONE, NONE, EIGHT, EIGHT, NONE,         // `elfie`
    ];

  const block = createBlock(str, entities);
  const encoded = encodeEntityRanges(block, {_6: '0', _8: '1'});
  expect(encoded).toMatchSnapshot();
});
