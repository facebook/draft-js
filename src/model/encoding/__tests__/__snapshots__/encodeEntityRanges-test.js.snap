// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`must handle ranges that include surrogate pairs 1`] = `
Array [
  Object {
    "key": "6",
    "length": 5,
    "offset": 6,
  },
  Object {
    "key": "8",
    "length": 2,
    "offset": 13,
  },
]
`;

exports[`must return an empty array if no entities present 1`] = `Array []`;

exports[`must return an empty array if no entities present 2`] = `Array []`;

exports[`must return interior ranges with multiple entities present 1`] = `
Array [
  Object {
    "key": "6",
    "length": 5,
    "offset": 2,
  },
  Object {
    "key": "8",
    "length": 1,
    "offset": 4,
  },
]
`;

exports[`must return overlapping ranges with multiple entities present 1`] = `
Array [
  Object {
    "key": "6",
    "length": 3,
    "offset": 2,
  },
  Object {
    "key": "8",
    "length": 3,
    "offset": 4,
  },
]
`;

exports[`must return ranges with an entity present more than once 1`] = `
Array [
  Object {
    "key": "6",
    "length": 2,
    "offset": 2,
  },
  Object {
    "key": "6",
    "length": 2,
    "offset": 5,
  },
]
`;

exports[`must return ranges with multiple entities present 1`] = `
Array [
  Object {
    "key": "6",
    "length": 2,
    "offset": 2,
  },
  Object {
    "key": "8",
    "length": 2,
    "offset": 5,
  },
]
`;

exports[`must return ranges with the storage-mapped key 1`] = `
Array [
  Object {
    "key": "6",
    "length": 2,
    "offset": 2,
  },
]
`;
